python/pyspark/sql/types.py

```diff
@@ -187,8 +187,12 @@ def needConversion(self):
 
     def toInternal(self, dt):
         if dt is not None:
-            seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
-                       else time.mktime(dt.timetuple()))
+            # Avoiding the invalid range of years (100-1899) for mktime in Python < 3
+            if dt.year > 1899 or dt.year <= 100:
+                seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
+                           else time.mktime(dt.timetuple()))
+            else:
+                seconds = calendar.timegm(dt.utctimetuple())
```

Comment:
I guess if `dt.tzinfo` doesn't exists, `dt` loses local timezone offset, so we need to adjust timestamp using `time.timezone` or something.